io/include/pcl/io/impl/pcd_io.hpp

```diff
@@ -105,19 +105,57 @@ pcl::PCDWriter::generateHeader (const pcl::PointCloud<PointT> &cloud, const int
   return (oss.str ());
 }
 
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointT> int
+pcl::PCDWriter::checkCloudSize (const pcl::PointCloud<PointT> &cloud)
+{
+  if (cloud.points.empty ())
+  {
+    throw pcl::IOException ("[pcl::PCDWriter::checkCloudSize] Input point cloud has no data or empty indices given!");
+    return (1);
+  }
+
+  if (cloud.width * cloud.height != cloud.points.size ())
+  {
+    PCL_WARN ("[pcl::PCDWriter::checkCloudSize]  Number of points different than width * height. File saved with width set to number of points");
+    return (2);
+  }
+  return (0);
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointT> int
+pcl::PCDWriter::checkCloudSize (const pcl::PointCloud<PointT> &cloud,
+                const std::vector<int> &indices)
+{
+  if (indices.empty ())
+  {
+    throw pcl::IOException ("[pcl::PCDWriter::checkCloudSize] Input point cloud has no data or empty indices given!");
+    return (1);
+  }
+
+  int res = checkCloudSize<PointT> (cloud);
+  if (res == 2)
+    return (2);
+  else if (res == 1)
+    return (1);
+  return (0);
+}
+
 /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 template <typename PointT> int
 pcl::PCDWriter::writeBinary (const std::string &file_name, 
                              const pcl::PointCloud<PointT> &cloud)
 {
-  if (cloud.empty ())
-  {
-    throw pcl::IOException ("[pcl::PCDWriter::writeBinary] Input point cloud has no data!");
+  int res = checkCloudSize<PointT> (cloud);
```

Comment:
Maybe use an enum, because keeping track of numbers is a bit dangerous.

Or return a struct with members `cloudOk` and `sizeOk` (sample names only).